### PR TITLE
fixing namespace for relationships and making it backwards compatible…

### PR DESCRIPTION
**Pull request by martafullen**
… (#690)

### CODE REVIEW

**Review by AlexanderWollbrink — APPROVED** [review]
my fix regarding the relationship namespaces, that falsely didn't get taken over to Eclipse. Thank you for cherry picking @martafullen

File: src/AasxCsharpLibrary/AdminShellPackageEnv.cs [hunk 1]
```diff
@@ -394,11 +394,25 @@ private static (AasCore.Aas3_0.Environment, Package) LoadPackageAasx(string fn,
                     }
 
                 if (originPart == null)
-                    throw (new Exception("Unable to find AASX origin. Aborting!"));
+                    xs = package.GetRelationshipsByType(
+                        "http://www.admin-shell.io/aasx/relationships/aasx-origin");
+                    foreach (var x in xs)
+                        if (x.SourceUri.ToString() == "/")
```
Comment (juileetikekar) [inline src/AasxCsharpLibrary/AdminShellPackageEnv.cs:400]: Why is here www being included?

File: src/AasxCsharpLibrary/AdminShellPackageEnv.cs [hunk 2]
```diff
@@ -411,7 +425,19 @@ private static (AasCore.Aas3_0.Environment, Package) LoadPackageAasx(string fn,
                 }
 
                 if (specPart == null)
-                    throw (new Exception("Unable to find AASX spec(s). Aborting!"));
+                    xs = originPart.GetRelationshipsByType("http://www.admin-shell.io/aasx/relationships/aas-spec");
+                    foreach (var x in xs)
```
Comment (juileetikekar) [inline src/AasxCsharpLibrary/AdminShellPackageEnv.cs:429]: Why is here "www" being included?

File: src/AasxCsharpLibrary/AdminShellPackageEnv.cs [hunk 3]
```diff
@@ -752,6 +778,26 @@ public enum SerializationFormat { None, Xml, Json };
                     PackagePart originPart = null;
                     var xs = package.GetRelationshipsByType(
                         "http://www.admin-shell.io/aasx/relationships/aasx-origin");
+                    foreach (var x in xs)
+                        if (x.SourceUri.ToString() == "/")
+                        {
+                            //originPart = package.GetPart(x.TargetUri);
+                            
+                            var absoluteURI = PackUriHelper.ResolvePartUri(x.SourceUri, x.TargetUri);
+                            if (package.PartExists(absoluteURI))
```
Comment (juileetikekar) [inline src/AasxCsharpLibrary/AdminShellPackageEnv.cs:787]: at line 780, i can see "www"

**Review by juileetikekar — COMMENTED** [review]
In addition to my comments, 

I could find one more place for a namespace change:
 "http://www.admin-shell.io/aasx/relationships/aas-suppl"

Shouldn't these also be changed?

**Comment by AlexanderWollbrink** [comment]
The fix for the bug is up.
I also added backwards compatible reading for "http://www.admin-shell.io/aasx/relationships/aas-suppl". Unfortunately, I can not see where the relationship for this is created, so I wasn't able to "convert" it to the new namespace. Maybe you could look into this @juileetikekar?

**Comment by juileetikekar** [comment]
Hi @AlexanderWollbrink,

Thank you for the fix.

Regarding fix related to "aas-suppl" file, please find my observations below:
1. When creating a new AASX V3 file => A correct new namespace gets added.
2. When converting V2 -> V3 => the namespace does not change, i.e., it remains "http://www.admin-shell.io/aasx/relationships/aas-suppl" (see the screenshot below).

image

The relationship gets created in case of supplementary files. If the AASX package has a supplimentary file, the relationship is found under "/aasx/<folder with name as AAS_Id>/_rels/<AAS_id>.aas.xml.rel"

You can use the below AASX file to reproduce the issue. Please rename the from from .zip to .aasx.

[TestAAS_V2.zip](https://github.com/eclipse-aaspe/aaspe/files/14979316/TestAAS_V2.zip)

If you unzip this file, you will find "AssetAdministrationShell---0F98619F.aas.xml.rels" file, which consists of "aas_suppl" relationship under "\aasx\AssetAdministrationShell---0F98619F\_rels\".

Please let me know if you need more information.

**Comment by AlexanderWollbrink** [comment]
Hey @juileetikekar,
thank you for the answer. 
My approach would be, to find the "old" namespace relationships, delete those and replace them with the new ones.
The issue is, that I can't find where in the code the existing relationships for suppl. files appear or where inside the package-structure they are saved.

**Comment by juileetikekar** [comment]
Hi @AlexanderWollbrink,

In my opinion, we will need to add a code explicitely to handle the namespace change. The current implementation just takes care of "aas_suppl" relationship only in case of adding and deleting file in case of "save" operation.

**Comment by AlexanderWollbrink** [comment]
I tested this with several AASX files that I had and it worked nicely for all the relationships.
In my opinion we can merge this into main.

File: src/AasxCsharpLibrary/AdminShellPackageEnv.cs [hunk 4]
```diff
@@ -394,11 +394,26 @@ private static (AasCore.Aas3_0.Environment, Package) LoadPackageAasx(string fn,
                     }
 
                 if (originPart == null)
-                    throw (new Exception("Unable to find AASX origin. Aborting!"));
+                    xs = package.GetRelationshipsByType(
+                        "http://www.admin-shell.io/aasx/relationships/aasx-origin");
+                    foreach (var x in xs)
+                        if (x.SourceUri.ToString() == "/")
```
Comment (Freezor) [inline src/AasxCsharpLibrary/AdminShellPackageEnv.cs:400]: I strongly recommend to use the **Equals** comparison method for strings.
```suggestion
                        if (x.SourceUri.Equals("/"))
```


- `==` Operator: The `==` operator compares the references of the string objects. If two strings have the same sequence of characters, but they are stored in different memory locations.
- `Equals()` Method: The `Equals()` method, on the other hand, compares the actual contents of the strings. It checks whether the sequences of characters in both strings are identical.

**Review by Freezor — APPROVED** [review]
In general the changes look fine to me. There are some smaller suggestions from my side. There still would be a lot to improve, but that could also mean to refactor that whole class.
I recommend, that we create unit tests for this class in the near future. That would really help us.

**Comment by mristin** [comment]
Just a heads-up: there is a unit tested & battle tested implementation of AASX package: https://github.com/aas-core-works/aas-package3-csharp

You might want to consider using it in the future.

**Review by juileetikekar — APPROVED** [review]

**Comment by juileetikekar** [comment]
My suggestion at this point is to merge the PR into main to unblock other activities. We can continue with the refactoring in next step.

@AlexanderWollbrink could you please merge the PR into main?

**Pull request merged**